### PR TITLE
Check for --complete-platforms match when --resolve-local-platforms

### LINKED ISSUE #1899: The `--resolve-local-platforms` option does not work with `--complete-platforms`
The `--resolve-local-platforms` option allows Pex to substitute a local interpreter for a `--platform` whenever there is a local interpreter available that fits the abbreviated `--platform` description. The option also applies to `--complete-platform` and in either case the test of a match is here:
https://github.com/pantsbuild/pex/blob/b1343ffbc4ef58c456d072c4f700ad23f1f51743/pex/resolve/target_configuration.py#L139-L161

The test is that there is an intersection between the local interpreter's compatibility tags and those of the foreign platform. Intersection is problematic when the local interpreter supports newer ABIs than the foreign platform. In this case, the local interpreter is used as a substitute for the foreign platform which allows wheels with newer ABIs than the foreign platform can support to be resolved. 

To correct this problem, a local interpreter should only be substituted for a foreign platform if its compatibility tags form a subset of the foreign platform tags. In the case of abbreviated `--platform` there is not much that can be done differently than the current intersection test since the full accurate set of tags is unknown. For the `--complete-platform` case, however, a subset relationship can be checked instead.

### PR DESCRIPTION
**Pull request by huonw**
This makes --resolve-local-platforms work better with --complete-platforms: if there's an external complete platform specified for a platform that matches a local interpreter, the local interpreter's complete platform is checked against that external complete platform. If the local interpreter supports any additional tags, the local interpreter is not used.

If the local interpreter/environment is newer (supports more tags), the old behaviour would risk including wheels that aren't compatible with the external complete platform. That is, execution will fail, see https://github.com/pantsbuild/pants/issues/17621 for an example.

Fixes #1899

### CODE REVIEW

**Review by jsirois — COMMENTED** [review]
Thanks for working on this @huonw.

I'm not fully convinced, but I think subset is the right choice still. The only reason to use `--resolve-local-platforms` at all is to make more resolves succeed by bringing sdists in play. If you want an "optimal" resolve and you have an environment that matches the target host, then you don't use `--platform` or `--complete-patform`, you just use `--python` to say this exact one right here - we have our ducks in a row and this matches production.

Maybe Pants does need to grow knobs as I think you suggested in another thread. Right now it tries to do the right thing with 0 knobs for the lambda case.

**Comment by huonw** [comment]
> I'm not fully convinced, but I think subset is the right choice still. The only reason to use --resolve-local-platforms at all is to make more resolves succeed by bringing sdists in play. If you want an "optimal" resolve and you have an environment that matches the target host, then you don't use --platform or --complete-patform, you just use --python to say this exact one right here - we have our ducks in a row and this matches production.

Ah, okay, and correspondingly, one can avoid using a "suboptimal" resolve due to a more limited local platform by not passing `--resolve-local-platforms` and only passing `--complete-platforms`. (The only hitch is users having the ability to disable `--resolve-local-platforms` for wrappers like `python_awslambda`.)

I think that makes sense, I'll switch to doing a subset check. I'm not sure of the exact check, but I suspect:

1. for `supported_tags`, doing `set(candidate_complete.supported_tags) <= set(requested_complete.supported_tags)` is enough? (Something like `all(requested_complete.supported_tags.rank(tag) is not None for tag in candidate_complete.supported_tags)` isn't required.)
2. `marker_environment` doesn't need to be checked (the 'interesting' parts are already captured by the `Platform` used in the dict for matching)

Does that sound correct? Thanks!

File: pex/resolve/target_configuration.py [hunk 1]
```diff
@@ -156,6 +156,25 @@ def resolve_targets(self):
                     )
                     if resolved_platforms:
                         for resolved_platform in resolved_platforms:
+                            # if there was an explicit complete platform specified, only use the
+                            # local interpreter if it exactly matches, or else incorrect wheels may
```
[the remark this answers is not in the document]
Comment (huonw) [reply pex/resolve/target_configuration.py:160]: Ah, thanks, I see what you mean: if the comment still applies after I switch to doing a subset computation, I'll reword it to something like "wheels with less preferred tags" to match the language of PEP 425.

**Comment by jsirois** [comment]
I think you've got that all right @huonw.

**Comment by huonw** [comment]
I've made that update, added a test, and also realised that the previous approach was fundamentally broken (I think an interpreter needs to actually be rejected if any of its intersecting platforms are incompatible, while the previous version continued to accepted it).

**Comment by jsirois** [comment]
I won't have time to dive deep until later, but Pex supports Python 2.7 still, so watch that syntax.

**Comment by huonw** [comment]
No worries.

Thanks for the 2.7 tip, I'll fix up the various errors (tomorrow, probably).

**Review by jsirois — APPROVED** [review]
Thanks @huonw - this looks good. I'll merge and get this out in a release shortly. You should be able to configure the `version` and [`known_versions`](https://www.pantsbuild.org/docs/reference-pex-cli#known_versions) `[pex-cli]` subsystem options in your Pants config to then be up and running with this at that point.

**Comment by huonw** [comment]
Woohoo, thank you!
